### ISSUE TITLE
[Umbrella] Enhance the engine service ha framework

### ISSUE THREAD

**Issue by timothy65535** [apache/incubator-kyuubi#918]
## Description

From the ha discovery design of engine service in codebase, there're some area that can be improved.

### 1. EngineRef is exposed to the KyuubiSessionImpl

When client connect to kyuubi server, the backendservice will open KyuubiSession. KyuubiSessionImpl use `EngineRef` and `ServiceDiscovery` at the same time in the codebase. 

The responsibility of `EngineRef` is to maintain the startup of the engine, it's better to move `EngineRef` to `EngineServiceDiscovery`. 
![image](https://user-images.githubusercontent.com/86483005/128994238-f1f6856b-9c41-4f6c-b844-829ad7052ee8.png)

### 2. The mechanism for determining whether engine started successfully is flawed

EngineRef use exists znode to determine whether engine started successfully or not(use the last znode), there are inconsistencies in this.
So, take an idea to improve the detection mechanism by adding the started sessionid to the contents of the znone when engine started, then check the sessionid.
![image](https://user-images.githubusercontent.com/86483005/128994276-5c9c7839-0a21-4253-8989-110b88ac0e74.png)

### 3. Missing support multiple engines under `USER` share level

Kyuubi supports `CONNECTION`, `USER`, `SERVER` share levels, we want to use kyuubi as hiveserver which can support multiple servers under same zookeeper namespace to share load pressure, client can pick hiveserver randomly. However, due to the influence of the second point above, kyuubi can not create other new engine under same user. Thes is also a little redundant to `subDomain`.
![image](https://user-images.githubusercontent.com/86483005/128994314-c978bf61-6a8f-4d95-9ea6-9fb96ca786cc.png)

### 4. Additional context

The first and second points can be optimized step by step, the third point requires more thought, welcome to discuss.

**Comment by ulysses-you** [comment]
thank you @timothy65535 for working on this.

About the second point. What if one user create two sessions, shall we reuse the same engine ? It sounds like the idea about engine pool that user can decide how many engine can be used in parallel.

**Comment by timothy65535** [comment]
Welcome @ulysses-you, the second point focus on improve the detection mechanism for engine startup. 

> What if one user create two sessions, shall we reuse the same engine

In my view, user still cannot create a second one unless the third point is completed, or meet concurrent bug.

The engine pool may needs to be implemented based on it. The third point a bit sames to https://github.com/apache/incubator-kyuubi/issues/786

**Comment by yaooqinn** [comment]
Maybe the names `[Kyuubi\Engine]ServiceDiscovery` are missleading, they are not used for discovering services but exposing services

the `object ServiceDiscovery` actually does the discovery jobs

**Comment by yaooqinn** [comment]
> 2. The mechanism for determining whether engine started successfully is flawed

I am thinking of adding an RPC that let an engine reporting its statuses during the initializing period

**Comment by timothy65535** [comment]
> Maybe the names `[Kyuubi\Engine]ServiceDiscovery` are misleading, they are not used for discovering services but exposing services

`EngineServiceDiscovery` just expose service itselft. `KyuubiServiceDiscovery` does discovery job.

> I am thinking of adding an RPC that let an engine reporting its statuses during the initializing period

Maybe is a new way to do. I think it's better to use zookeeper, it can help us decouple KyuubiServer and SparkEngine.

**Comment by yaooqinn** [comment]
KyuubiServiceDiscovery does not

**Comment by timothy65535** [comment]
> KyuubiServiceDiscovery does not

Yeah, the names [Kyuubi\Engine]ServiceDiscovery are misleading. 😂

**Comment by yaooqinn** [comment]
Maybe we can rename `[Kyuubi\Engine]ServiceRegister `

**Comment by timothy65535** [comment]
We need to think it in deep, take the `hive` project as an example, seems using discovery is ok.

**Comment by timothy65535** [comment]
Thanks share your idea. Conceptually, ServiceDiscovery contains `register` and `discovery` two modules. In the codebase, most of the module designs of`ServiceDiscovery`, `KyuubiServiceDiscovery`, `EngineServiceDiscovery` are ok, maybe just a little change. 

So, have an idea that we can minimize the issue scope and do these separately. We can start with the first and second points, they will not affect the following future change.

**Comment by yaooqinn** [comment]
Please send a PR for option 2 first？

**Comment by timothy65535** [comment]
> Please send a PR for option 2 first？

OK

**Comment by timothy65535** [comment]
- [ ] Extract the zookeeper part from ServiceDiscovery
